From 0af3ae2d8b31e31c991eb7dd1d4b79e698907c8d Mon Sep 17 00:00:00 2001
From: Vladimir Serbinenko <phcoder@gmail.com>
Date: Sun, 10 Aug 2014 11:27:13 +0200
Subject: [PATCH 091/506] 	* grub-core/disk/i386/pc/biosdisk.c
 (grub_biosdisk_rw): Add 	safety to avoid triggerring VirtualBox bug.

---
 ChangeLog                         | 5 +++++
 grub-core/disk/i386/pc/biosdisk.c | 8 ++++++++
 2 files changed, 13 insertions(+)

diff --git a/ChangeLog b/ChangeLog
index cab1427..9e159a8 100644
--- a/ChangeLog
+++ b/ChangeLog
@@ -1,5 +1,10 @@
 2014-08-10  Vladimir Serbinenko  <phcoder@gmail.com>
 
+	* grub-core/disk/i386/pc/biosdisk.c (grub_biosdisk_rw): Add
+	safety to avoid triggerring VirtualBox bug.
+
+2014-08-10  Vladimir Serbinenko  <phcoder@gmail.com>
+
 	* grub-core/fs/cbfs.c: Don't probe disks of unknow size.
 
 	Fixes hang on virtualbox.
diff --git a/grub-core/disk/i386/pc/biosdisk.c b/grub-core/disk/i386/pc/biosdisk.c
index 1689539..6b21525 100644
--- a/grub-core/disk/i386/pc/biosdisk.c
+++ b/grub-core/disk/i386/pc/biosdisk.c
@@ -455,6 +455,14 @@ grub_biosdisk_rw (int cmd, grub_disk_t disk,
 {
   struct grub_biosdisk_data *data = disk->data;
 
+  /* VirtualBox fails with sectors above 2T on CDs.
+     Since even BD-ROMS are never that big anyway, return error.  */
+  if ((data->flags & GRUB_BIOSDISK_FLAG_CDROM)
+      && (sector >> 32))
+    return grub_error (GRUB_ERR_OUT_OF_RANGE,
+		       N_("attempt to read or write outside of disk `%s'"),
+		       disk->name);
+
   if (data->flags & GRUB_BIOSDISK_FLAG_LBA)
     {
       struct grub_biosdisk_dap *dap;
-- 
2.4.3

